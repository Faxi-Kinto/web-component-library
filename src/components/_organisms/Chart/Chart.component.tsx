import React from 'react';
import ReactApexChart from 'react-apexcharts';
import moment from 'moment';

export type ChartSeries = {
  name: string;
  data: (number | any)[];
};

export type TitleStyle = {
  fontSize?: string;
  fontWeight?: string;
  fontFamily?: string;
  color?: string;
};

export type ChartProps = {
  series: ChartSeries[];
  chartHeight: number | string;
  colors: string[]; // line colors
  gradientToColors: string[];
  title: string;
  chartType: 'line'; // limit types here if more types available
  strokeWidths: number[];
  xAxisCssClass?: string;
  yAxisCssClass?: string;
  titleStyle?: TitleStyle;
  xAxisTickCount?: number;
  xAxisMinMiliseconds: number;
  xAxisMaxMiliseconds: number;
  markerStrokeColors: string[];
};

const Chart: React.FC<ChartProps> = (props: ChartProps): JSX.Element => {
  const {
    title,
    chartHeight,
    chartType,
    series,
    colors,
    strokeWidths,
    gradientToColors,
    xAxisCssClass = '',
    yAxisCssClass = '',
    titleStyle = {},
    xAxisTickCount = 2,
    xAxisMinMiliseconds,
    xAxisMaxMiliseconds,
    markerStrokeColors,
  } = props;

  const state = {
    options: {
      title: {
        text: title,
        align: 'left',
        style: titleStyle,
      },
      chart: {
        height: chartHeight,
        type: chartType,
        toolbar: {
          show: false,
        },
        zoom: {
          enabled: false,
        },
      },
      colors,
      fill: {
        type: 'gradient',
        gradient: {
          shadeIntensity: 1,
          inverseColors: false,
          opacityFrom: 1,
          opacityTo: 1,
          gradientToColors,
          stops: [20, 100, 100, 100],
        },
      },
      grid: {
        xaxis: {
          lines: {
            show: true,
          },
        },
        yaxis: {
          lines: {
            show: true,
          },
        },
        borderColor: '#C6CDCF',
        strokeDashArray: 1,
      },
      markers: {
        size: 3,
        colors: ['white'],
        strokeColors: markerStrokeColors,
        strokeWidth: 2,
        hover: {
          size: 4,
        },
      },
      series,
      stroke: {
        width: strokeWidths,
        curve: 'smooth',
      },
      yaxis: {
        labels: {
          offsetX: 0,
          formatter: (num: number) => {
            return num.toFixed(0);
          },
          style: {
            cssClass: yAxisCssClass,
          },
        },
        axisBorder: {
          show: true,
          color: '#C6CDCF',
          width: 1,
        },
      },
      xaxis: {
        type: 'datetime',
        tickAmount: xAxisTickCount,
        min: xAxisMinMiliseconds,
        max: xAxisMaxMiliseconds,
        labels: {
          rotateAlways: true,
          formatter: (_: any, timestamp: string | number | Date) => {
            return moment(new Date(timestamp)).format('DD MMM');
          },
          style: {
            cssClass: xAxisCssClass,
          },
        },
        axisBorder: {
          show: true,
          color: '#C6CDCF',
          height: 1,
        },
        tooltip: {
          enabled: false,
        },
      },
      legend: {
        show: true,
        markers: {
          width: 35,
          height: 2,
          radius: 10,
        },
      },
    },
  };

  return (
    <ReactApexChart options={state.options} series={state.options.series} />
  );
};

export default Chart;
