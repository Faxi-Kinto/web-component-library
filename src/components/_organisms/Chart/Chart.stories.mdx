import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { useState } from 'react';
import Chart from './Chart.component.tsx';

<Meta title="Core|Organisms/Chart" />

# Chart Component

This is an implementation of the Chart Component

## Implementation

#### Single data series

<Preview>
  <Story name="single data series">
    <div style={{height: '500px', width: '500px' }}>
    <Chart
      title="Number of messages"
      chartHeight="100%"
      chartType="line"
      series={[
        {
          data: [
            { x: new Date(1593993600000), y: 1 },
            { x: new Date(1594080000000), y: 1 },
            { x: new Date(1594166400000), y: 0 },
            { x: new Date(1594252800000), y: 0 },
            { x: new Date(1594339200000), y: 1 },
            { x: new Date(1594425600000), y: 0 },
            { x: new Date(1594512000000), y: 0 },
            { x: new Date(1594598400000), y: 0 },
            { x: new Date(1594684800000), y: 0 },
            { x: new Date(1594771200000), y: 1 },
            { x: new Date(1594857600000), y: 1 },
            { x: new Date(1594944000000), y: 0 },
            { x: new Date(1595030400000), y: 1 },
            { x: new Date(1595116800000), y: 18 },
            { x: new Date(1595203200000), y: 1 },
            { x: new Date(1595289600000), y: 0 },
            { x: new Date(1595376000000), y: 1 },
            { x: new Date(1595462400000), y: 1 },
            { x: new Date(1595548800000), y: 0 },
            { x: new Date(1595635200000), y: 0 },
            { x: new Date(1595721600000), y: 1 },
            { x: new Date(1595808000000), y: 0 },
            { x: new Date(1595894400000), y: 0 },
            { x: new Date(1595980800000), y: 0 },
            { x: new Date(1596067200000), y: 0 },
            { x: new Date(1596153600000), y: 0 },
            { x: new Date(1596240000000), y: 0 },
            { x: new Date(1596326400000), y: 0 },
            { x: new Date(1596412800000), y: 0 },
            { x: new Date(1596499200000), y: 0 },
            { x: new Date(1596585600000), y: 1 },
            { x: new Date(1596672000000), y: 1 },
          ],
          name: 'Messages',
        },
      ]}
      colors={['#81C8D1']}
      gradientToColors={['#00708D']}
      strokeWidths={[2]}
      xAxisMinMiliseconds={new Date(1593993600000).getTime()}
      xAxisMaxMiliseconds={new Date(1596672000000).getTime()}
      markerStrokeColors={['#00A5B5']}
      yAxisTickInterval={3}
    />
    </div>
  </Story>
</Preview>

#### Multiple data series

<Preview>
  <Story name="multiple data series">
    <div style={{height: '500px', width: '500px' }}>
      <Chart
        title="Number of users"
        chartHeight="100%"
        chartType="line"
        series={[
          {
            data: [
              {
                x: new Date('2020-06-05').getTime(),
                y: 20,
              },
              {
                x: new Date('2020-06-15').getTime(),
                y: 50,
              },
              {
                x: new Date('2020-06-30').getTime(),
                y: 70,
              },
            ],
            name: 'Italy',
          },
                {
            data: [
              {
                x: new Date('2020-06-07').getTime(),
                y: 15,
              },
              {
                x: new Date('2020-06-17').getTime(),
                y: 60,
              },
              {
                x: new Date('2020-06-27').getTime(),
                y: 100,
              },
            ],
            name: 'Serbia',
          },
        ]}
        colors={['#81C8D1', '#C6CDCF']}
        gradientToColors={['#00708D', '#4E606A']}
        strokeWidths={[2, 2]}
        xAxisMinMiliseconds={new Date('06/01/2020').getTime()}
        xAxisMaxMiliseconds={new Date('07/01/2020').getTime()}
        markerStrokeColors={['#00A5B5', '#4E606A']}
      />
    </div>
  </Story>
</Preview>

# Props

<Props of={Chart} />
