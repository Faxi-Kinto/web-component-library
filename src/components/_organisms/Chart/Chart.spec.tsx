import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Chart from './Chart.component';

test('Chart renders correctly', () => {
  const renderer = ShallowRenderer.createRenderer();

  const component = renderer.render(
    <Chart
      title="Number of messages"
      height="100%"
      chartType="line"
      series={[
        {
          data: [
            { x: new Date(1593993600000).getTime(), y: '2' },
            { x: new Date(1594080000000).getTime(), y: '2' },
            { x: new Date(1594166400000).getTime(), y: '0' },
            { x: new Date(1594252800000).getTime(), y: '0' },
            { x: new Date(1594339200000).getTime(), y: '1' },
            { x: new Date(1594425600000).getTime(), y: '0' },
            { x: new Date(1594512000000).getTime(), y: '0' },
            { x: new Date(1594598400000).getTime(), y: '0' },
            { x: new Date(1594684800000).getTime(), y: '0' },
            { x: new Date(1594771200000).getTime(), y: '0' },
            { x: new Date(1594857600000).getTime(), y: '0' },
            { x: new Date(1594944000000).getTime(), y: '0' },
            { x: new Date(1595030400000).getTime(), y: '0' },
            { x: new Date(1595116800000).getTime(), y: '0' },
            { x: new Date(1595203200000).getTime(), y: '0' },
            { x: new Date(1595289600000).getTime(), y: '0' },
            { x: new Date(1595376000000).getTime(), y: '1' },
            { x: new Date(1595462400000).getTime(), y: '0' },
            { x: new Date(1595548800000).getTime(), y: '0' },
            { x: new Date(1595635200000).getTime(), y: '0' },
            { x: new Date(1595721600000).getTime(), y: '0' },
            { x: new Date(1595808000000).getTime(), y: '0' },
            { x: new Date(1595894400000).getTime(), y: '0' },
            { x: new Date(1595980800000).getTime(), y: '0' },
            { x: new Date(1596067200000).getTime(), y: '0' },
            { x: new Date(1596153600000).getTime(), y: '0' },
            { x: new Date(1596240000000).getTime(), y: '0' },
            { x: new Date(1596326400000).getTime(), y: '0' },
            { x: new Date(1596412800000).getTime(), y: '0' },
            { x: new Date(1596499200000).getTime(), y: '0' },
            { x: new Date(1596585600000).getTime(), y: '2' },
            { x: new Date(1596672000000).getTime(), y: '1' },
          ],
          name: 'Messages',
        },
      ]}
      colors={['#81C8D1']}
      gradientToColors={['#00708D']}
      strokeWidths={[2]}
      xAxisMinMiliseconds={new Date(1593993600000).getTime()}
      xAxisMaxMiliseconds={new Date(1596672000000).getTime()}
      markerStrokeColors={['#00A5B5']}
      yAxisTickInterval={2}
    />
  );

  expect(component).toMatchSnapshot();
});
