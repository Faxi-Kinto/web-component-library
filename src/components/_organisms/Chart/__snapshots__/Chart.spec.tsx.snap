// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart renders correctly 1`] = `
<r
  height="100%"
  options={
    Object {
      "chart": Object {
        "toolbar": Object {
          "show": false,
        },
        "type": "line",
        "zoom": Object {
          "enabled": false,
        },
      },
      "colors": Array [
        "#81C8D1",
      ],
      "fill": Object {
        "gradient": Object {
          "gradientToColors": Array [
            "#00708D",
          ],
          "inverseColors": false,
          "opacityFrom": 1,
          "opacityTo": 1,
          "shadeIntensity": 1,
          "stops": Array [
            20,
            100,
            100,
            100,
          ],
        },
        "type": "gradient",
      },
      "grid": Object {
        "borderColor": "#C6CDCF",
        "strokeDashArray": 1,
        "xaxis": Object {
          "lines": Object {
            "show": true,
          },
        },
        "yaxis": Object {
          "lines": Object {
            "show": true,
          },
        },
      },
      "legend": Object {
        "markers": Object {
          "height": 2,
          "radius": 10,
          "width": 35,
        },
        "show": true,
      },
      "markers": Object {
        "colors": Array [
          "white",
        ],
        "hover": Object {
          "size": 4,
        },
        "size": 3,
        "strokeColors": Array [
          "#00A5B5",
        ],
        "strokeWidth": 2,
      },
      "series": Array [
        Object {
          "data": Array [
            Object {
              "x": 1593993600000,
              "y": "2",
            },
            Object {
              "x": 1594080000000,
              "y": "2",
            },
            Object {
              "x": 1594166400000,
              "y": "0",
            },
            Object {
              "x": 1594252800000,
              "y": "0",
            },
            Object {
              "x": 1594339200000,
              "y": "1",
            },
            Object {
              "x": 1594425600000,
              "y": "0",
            },
            Object {
              "x": 1594512000000,
              "y": "0",
            },
            Object {
              "x": 1594598400000,
              "y": "0",
            },
            Object {
              "x": 1594684800000,
              "y": "0",
            },
            Object {
              "x": 1594771200000,
              "y": "0",
            },
            Object {
              "x": 1594857600000,
              "y": "0",
            },
            Object {
              "x": 1594944000000,
              "y": "0",
            },
            Object {
              "x": 1595030400000,
              "y": "0",
            },
            Object {
              "x": 1595116800000,
              "y": "0",
            },
            Object {
              "x": 1595203200000,
              "y": "0",
            },
            Object {
              "x": 1595289600000,
              "y": "0",
            },
            Object {
              "x": 1595376000000,
              "y": "1",
            },
            Object {
              "x": 1595462400000,
              "y": "0",
            },
            Object {
              "x": 1595548800000,
              "y": "0",
            },
            Object {
              "x": 1595635200000,
              "y": "0",
            },
            Object {
              "x": 1595721600000,
              "y": "0",
            },
            Object {
              "x": 1595808000000,
              "y": "0",
            },
            Object {
              "x": 1595894400000,
              "y": "0",
            },
            Object {
              "x": 1595980800000,
              "y": "0",
            },
            Object {
              "x": 1596067200000,
              "y": "0",
            },
            Object {
              "x": 1596153600000,
              "y": "0",
            },
            Object {
              "x": 1596240000000,
              "y": "0",
            },
            Object {
              "x": 1596326400000,
              "y": "0",
            },
            Object {
              "x": 1596412800000,
              "y": "0",
            },
            Object {
              "x": 1596499200000,
              "y": "0",
            },
            Object {
              "x": 1596585600000,
              "y": "2",
            },
            Object {
              "x": 1596672000000,
              "y": "1",
            },
          ],
          "name": "Messages",
        },
      ],
      "stroke": Object {
        "curve": "smooth",
        "width": Array [
          2,
        ],
      },
      "title": Object {
        "align": "left",
        "style": Object {},
        "text": "Number of messages",
      },
      "xaxis": Object {
        "axisBorder": Object {
          "color": "#C6CDCF",
          "height": 1,
          "show": true,
        },
        "labels": Object {
          "formatter": [Function],
          "rotateAlways": true,
          "style": Object {
            "cssClass": "",
          },
        },
        "max": 1596672000000,
        "min": 1593993600000,
        "tickAmount": 2,
        "tooltip": Object {
          "enabled": false,
        },
        "type": "datetime",
      },
      "yaxis": Object {
        "axisBorder": Object {
          "color": "#C6CDCF",
          "show": true,
          "width": 1,
        },
        "labels": Object {
          "formatter": [Function],
          "offsetX": 0,
          "style": Object {
            "cssClass": "",
          },
        },
        "tickAmount": 2,
      },
    }
  }
  series={
    Array [
      Object {
        "data": Array [
          Object {
            "x": 1593993600000,
            "y": "2",
          },
          Object {
            "x": 1594080000000,
            "y": "2",
          },
          Object {
            "x": 1594166400000,
            "y": "0",
          },
          Object {
            "x": 1594252800000,
            "y": "0",
          },
          Object {
            "x": 1594339200000,
            "y": "1",
          },
          Object {
            "x": 1594425600000,
            "y": "0",
          },
          Object {
            "x": 1594512000000,
            "y": "0",
          },
          Object {
            "x": 1594598400000,
            "y": "0",
          },
          Object {
            "x": 1594684800000,
            "y": "0",
          },
          Object {
            "x": 1594771200000,
            "y": "0",
          },
          Object {
            "x": 1594857600000,
            "y": "0",
          },
          Object {
            "x": 1594944000000,
            "y": "0",
          },
          Object {
            "x": 1595030400000,
            "y": "0",
          },
          Object {
            "x": 1595116800000,
            "y": "0",
          },
          Object {
            "x": 1595203200000,
            "y": "0",
          },
          Object {
            "x": 1595289600000,
            "y": "0",
          },
          Object {
            "x": 1595376000000,
            "y": "1",
          },
          Object {
            "x": 1595462400000,
            "y": "0",
          },
          Object {
            "x": 1595548800000,
            "y": "0",
          },
          Object {
            "x": 1595635200000,
            "y": "0",
          },
          Object {
            "x": 1595721600000,
            "y": "0",
          },
          Object {
            "x": 1595808000000,
            "y": "0",
          },
          Object {
            "x": 1595894400000,
            "y": "0",
          },
          Object {
            "x": 1595980800000,
            "y": "0",
          },
          Object {
            "x": 1596067200000,
            "y": "0",
          },
          Object {
            "x": 1596153600000,
            "y": "0",
          },
          Object {
            "x": 1596240000000,
            "y": "0",
          },
          Object {
            "x": 1596326400000,
            "y": "0",
          },
          Object {
            "x": 1596412800000,
            "y": "0",
          },
          Object {
            "x": 1596499200000,
            "y": "0",
          },
          Object {
            "x": 1596585600000,
            "y": "2",
          },
          Object {
            "x": 1596672000000,
            "y": "1",
          },
        ],
        "name": "Messages",
      },
    ]
  }
  type="line"
  width="100%"
/>
`;
